pkgs/by-name/im/immich/package.nix

```
@@ -0,0 +1,318 @@
+      mkdir -p $out
+      mv package.json package-lock.json node_modules dist $out/
+
+      ls $out/dist
+
+      makeWrapper ${lib.getExe nodejs} $out/bin/immich --add-flags $out/dist/index.js
+
+      runHook postInstall
+    '';
+
+    meta = {
+      inherit (meta)
+        description
+        homepage
+        license
+        maintainers
+        platforms
+        ;
+      mainProgram = "immich";
+    };
+  };
+
+  web = buildNpmPackage' {
+    pname = "immich-web";
+    inherit version;
+    src = "${src}/web";
+    inherit (sources.components.web) npmDepsHash;
+
+    inherit (cli) preBuild;
+
+    installPhase = ''
+      runHook preInstall
+
+      cp -r build $out
+
+      runHook postInstall
+    '';
+  };
+
+  machine-learning = python3.pkgs.buildPythonApplication {
+    pname = "immich-machine-learning";
+    inherit version;
+    src = "${src}/machine-learning";
+    pyproject = true;
+
+    postPatch = ''
+      substituteInPlace pyproject.toml --replace-fail 'fastapi-slim' 'fastapi'
+
+      # Allow immich to use pydantic v2
+      substituteInPlace app/schemas.py --replace-fail 'pydantic' 'pydantic.v1'
+      substituteInPlace app/main.py --replace-fail 'pydantic' 'pydantic.v1'
+      substituteInPlace app/config.py \
+        --replace-fail 'pydantic' 'pydantic.v1' \
+        --replace-fail '/cache' '/var/cache/immich'
+    '';
+
+    pythonRelaxDeps = [
+      "setuptools"
+      "pydantic"
+    ];
+    pythonRemoveDeps = [ "opencv-python-headless" ];
+
+    nativeBuildInputs = with python3.pkgs; [
+      pythonRelaxDepsHook
+      poetry-core
+      cython
+    ];
+
+    dependencies =
+      with python3.pkgs;
+      [
+        insightface
+        opencv4
+        pillow
+        fastapi
+        uvicorn
+        aiocache
+        rich
+        ftfy
+        setuptools
+        python-multipart
+        orjson
+        gunicorn
+        huggingface-hub
+        tokenizers
+        pydantic
+      ]
+      ++ uvicorn.optional-dependencies.standard;
+
+    doCheck = false;
+
+    postInstall = ''
+      mkdir -p $out/share
+      cp log_conf.json $out/share
+
+      cp -r ann $out/${python3.sitePackages}/
+
+      makeWrapper ${python3.pkgs.gunicorn}/bin/gunicorn $out/bin/machine-learning \
+        --prefix PYTHONPATH : "$out/${python3.sitePackages}:$PYTHONPATH" \
+        --set-default MACHINE_LEARNING_WORKERS 1 \
+        --set-default MACHINE_LEARNING_WORKER_TIMEOUT 120 \
+        --set-default IMMICH_HOST 127.0.0.1 \
+        --set-default IMMICH_PORT 3003 \
+        --add-flags "app.main:app -k app.config.CustomUvicornWorker \
+          -w \"\$MACHINE_LEARNING_WORKERS\" \
+          -b \"\$IMMICH_HOST:\$IMMICH_PORT\" \
+          -t \"\$MACHINE_LEARNING_WORKER_TIMEOUT\"
+          --log-config-json $out/share/log_conf.json"
+    '';
+
+    meta = {
+      inherit (meta)
+        description
+        homepage
+        license
+        maintainers
+        platforms
+        ;
+      mainProgram = "machine-learning";
```

```suggestion
    meta = meta // {
      mainProgram = "machine-learning";
    }
```
and the same in other places.

---

No, this is wrong, it will break meta.position for example.

---

This was also done in response to another review comment for `meta.position`

---

Sure, but does it really matter since it's in the same file in this case? It makes it so much more readable.

---

Are we really nitting here to encourage _bad_ practices?

---

I agree it's less correct.
Might I suggest to split this into multiple files then in order to improve readability?

---

Seems reasonable.

---

I do not see how more files would make it more readable. Now instead of jumping to definitions in just one file, you need to keep track of several files. Also this would be quite annoying to do as some intermediate derivations are shared between e.g. cli and server.

---

I don't see what's shared between cli and server.
I suggest we have a `machine-learning.nix` and `cli.nix` because those two expressions are pretty long by themselves.

---

the openapi package for example. The prebuild step as well

---

Those can easily be passed to the appropriate `import`.